src/rust/cpu/cpu.rs

```diff
@@ -1940,7 +1943,10 @@ pub unsafe fn do_page_walk(
             if user {
                 // Page Fault: page table accessed by non-supervisor
                 if side_effects {
-                    trigger_pagefault(addr, true, for_writing, user, jit);
+                    // NOTE: page presence has to be checked since we
+                    //       don't know about page table entry
+                    let is_present = is_page_present_in_page_directory(addr, page_dir_entry);
+                    trigger_pagefault(addr, is_present, for_writing, user, jit);
                 }
                 return Err(());
             }
```

Comment:
Change this entire block to `allow_user &= page_dir_entry & PAGE_TABLE_USER_MASK != 0;`

Then, in the page table entry rewrite the user check to:

```rust
allow_user &= page_table_entry & PAGE_TABLE_USER_MASK != 0;
if user && !allow_user {
    if side_effects {
        trigger_pagefault(addr, true, for_writing, user, jit);
    }
    return Err(());
}
```

The size bit set case also needs a similar check then.

The same should work for write checks (need to introduce a variable similar to `allow_user`).

Reply:
Thank you, I will do the changes soon.